src/ext/command.rs

```diff
@@ -0,0 +1,165 @@
+
+use std::{
+    fmt::Display,
+    process::{Command, Output},
+};
+
+use getset::Getters;
+
+/// The description of a command.
+#[derive(Debug, Clone, Getters)]
+#[getset(get = "pub")]
+pub struct CommandDescription {
+    /// The name of the command.
+    name: String,
+
+    /// The arguments to the command.
+    args: Vec<String>,
+
+    /// The environment variables provided to the command.
+    envs: Vec<String>,
+
+    /// The status code of the command, if any.
+    status: Option<i32>,
+
+    /// The stdout of the command, if any.
+    stdout: Option<String>,
+
+    /// The stderr of the command, if any.
+    stderr: Option<String>,
+}
+
+impl Display for CommandDescription {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "{}", self.name())?;
+        writeln!(f, "args: {}", self.display_args())?;
+        writeln!(f, "env: {}", self.display_envs())?;
+        if let Some(status) = &self.status {
+            writeln!(f, "status: {}", status)?;
+        }
+        if let Some(stdout) = &self.stdout {
+            writeln!(f, "stdout: '{}'", stdout.trim())?;
+        }
+        if let Some(stderr) = &self.stderr {
+            writeln!(f, "stderr: '{}'", stderr.trim())?;
+        }
+        Ok(())
+    }
+}
+
+impl CommandDescription {
+    /// Enrich the command description with the command's stdout.
+    pub fn with_stdout(self, stdout: String) -> Self {
+        Self {
+            stdout: Some(stdout),
+            ..self
+        }
+    }
+
+    /// Enrich the command description with the command's stderr.
+    pub fn with_stderr(self, stderr: String) -> Self {
+        Self {
+            stderr: Some(stderr),
+            ..self
+        }
+    }
+
+    /// Enrich the command description with the command's status code.
+    pub fn with_status(self, status: i32) -> Self {
+        Self {
+            status: Some(status),
+            ..self
+        }
+    }
+
+    /// Enrich the command description with the command output.
+    /// Convenience method for `with_stdout`, `with_stderr`, and `with_status` when the user has an `Output`.
+    pub fn with_output(self, output: &Output) -> Self {
+        let stdout = String::from_utf8_lossy(&output.stdout);
+        let stderr = String::from_utf8_lossy(&output.stderr);
+        Self {
+            stdout: Some(stdout.to_string()),
+            stderr: Some(stderr.to_string()),
+            status: Some(output.status.code().unwrap_or(-1)),
+            ..self
+        }
+    }
+
+    /// Joins the args into a single string for display.
+    ///
+    /// This is preferred over using the debug implementation because if the args contain a path,
+    /// and Broker is running on Windows, the debug implementation doubles backslashes.
+    pub fn display_args(&self) -> String {
+        let joined = self
+            .args
+            .iter()
+            .map(|a| format!(r#""{a}""#))
+            .collect::<Vec<_>>()
```

Comment:
I have not fully reviewed this yet, but I noticed that this can leak authentication info and wanted to raise it.

This didn't get introduced in this PR. It's just the first time I noticed it, and I think some of the ideas in this file would make it easier to implement a fix.

The repro case is just to add an http basic_auth or header integration that does not work. E.g.

```yaml
  - type: git
    poll_interval: 1h
    remote: https://gitlab.com/scottpatten/serially-server
    auth:
      type: http_basic
      username: scottpatten
      password: glpat-not-a-real-token
```

When you run `cargo run -- run`, you see:

```
  2023-03-25T00:04:30.255254Z  WARN broker::subcommand::run: attempt to poll integration at https://gitlab.com/scottpatten/serially-server failed: run external command: IO layer error: run command: git
args: ["-c", "credential.helper=", "-c", "http.extraHeader=AUTHORIZATION: Basic c2NvdHRwYXR0ZW46Z2xwYXQtbm90LWEtcmVhbC10b2tlbg==", "ls-remote", "--quiet"]
env: ["GCM_INTERACTIVE=never", "GIT_ASKPASS=<REMOVED>", "GIT_TERMINAL_PROMPT=0"]
status: 128
stdout: ''
stderr: 'fatal: could not read Username for 'https://gitlab.com': terminal prompts disabled'
```

That "c2NvdHRwYXR0ZW46Z2xwYXQtbm90LWEtcmVhbC10b2tlbg==" is just `username:password` base64 encoded.

```
echo c2NvdHRwYXR0ZW46Z2xwYXQtbm90LWEtcmVhbC10b2tlbg== | base64 --decode
scottpatten:glpat-not-a-real-token
```

I'm not 100% sure of the best way to fix this, but one idea is to make a wrapper around `Command::new` that takes in args that are either String or ComparableSecretString, and then display knows whether they are secrets or not

Reply:
🤔 hmm, good catch! I'll make a follow up PR to address this.

Reply:
Sounds great! We definitely don't need to fix it in this PR